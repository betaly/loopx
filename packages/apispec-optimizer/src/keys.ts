import {BindingKey} from '@loopback/core';

import {ApiSpecOptimizerOptions} from './optimizer.spec-enhancer';

export namespace ApiSpecOptimizerBindings {
  /**
   * Strongly-typed binding key for ApiSpecOptimizer
   */
  export const API_SPEC_OPTIMIZER = BindingKey.create<ApiSpecOptimizerOptions>('oas-enhancer.ApiSpecOptimizer');
}
