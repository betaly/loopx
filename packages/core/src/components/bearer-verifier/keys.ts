import {BindingKey} from '@loopback/context';

import {BINDING_PREFIX} from '../../constants';

export namespace BearerVerifierBindings {
  export const Config = BindingKey.create<BearerVerifierConfig>(`${BINDING_PREFIX}.bearer-verfier.config`);
}

export enum BearerVerifierType {
  service,
  facade,
}

export interface BearerVerifierConfig {
  type: BearerVerifierType;
}

// export interface IAuthUserWithPermissions<ID = string, TID = string, UTID = string>
//   extends MarkOptional<OmitIndexSignature<IAuthTenantUser<ID, TID, UTID>>, 'id'> {
//   permissions: string[];
//
//   // eslint-disable-next-line @typescript-eslint/no-explicit-any
//   [attribute: string]: any;
// }
