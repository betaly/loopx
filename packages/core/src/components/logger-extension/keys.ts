import {BindingKey} from '@loopback/context';

import {BINDING_PREFIX} from '../../constants';
import {ILogger} from './logger.interface';

export namespace LOGGER {
  /**
   * Injection key constant
   */
  export const LOGGER_INJECT = `${BINDING_PREFIX}.log.action`;

  /**
   * Binding keys used by this component.
   */
  export namespace BINDINGS {
    export const LOG_ACTION = BindingKey.create<ILogger>(LOGGER_INJECT);
  }

  /**
   * Enum to define the supported log levels
   */
  export enum LOG_LEVEL {
    DEBUG,
    INFO,
    WARN,
    ERROR,
    OFF,
  }
}
