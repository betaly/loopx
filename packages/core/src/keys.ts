import {BindingKey} from '@loopback/core';
import {ExpressRequestHandler} from '@loopback/rest';

import {BINDING_PREFIX} from './constants';
import {HttpMethod} from './enums';
import {CoreConfig} from './types';

export namespace LxCoreBindings {
  export const i18n = BindingKey.create<i18nAPI>(`${BINDING_PREFIX}.i18n`);

  export const config = BindingKey.create<CoreConfig>('loopx.packages.core.config');

  export const EXPRESS_MIDDLEWARES = BindingKey.create<ExpressRequestHandler[]>(
    `loopx.packages.core.expressMiddlewares`,
  );

  export const DEFAULT_TENANT_KEY = BindingKey.create<string>('loopx.packages.core.defaultTenantKey');
}

const hiddenKey = 'loopx.oas.hiddenEndpoints';

export type OasHiddenApi = {
  path: string;
  httpMethod: HttpMethod;
};

export namespace OASBindings {
  export const HiddenEndpoint = BindingKey.create<OasHiddenApi[]>(hiddenKey);
}
