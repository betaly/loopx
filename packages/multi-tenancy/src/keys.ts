import {BindingKey} from '@loopback/core';

import {IdentifyTenantFn, MultiTenancyConfig, MultiTenancyPostProcess, Tenant, TenantResolverFn} from './types';

export namespace MultiTenancyBindings {
  export const CONFIG = BindingKey.create<MultiTenancyConfig>('loopx.multi-tenancy.config');

  export const ACTION = BindingKey.create<IdentifyTenantFn>('loopx.multi-tenancy.identifyTenant');

  export const CURRENT_TENANT = BindingKey.create<Tenant>('loopx.multi-tenancy.currentTenant');

  export const DEFAULT_TENANT_ID = BindingKey.create<string>('loopx.multi-tenancy.defaultTenantId');

  export const POST_PROCESS = BindingKey.create<MultiTenancyPostProcess>('loopx.multi-tenancy.postProcess');

  export const TENANT_RESOLVER = BindingKey.create<TenantResolverFn>('loopx.multi-tenancy.tenantResolver');
}

export const MULTI_TENANCY_STRATEGIES = 'loopx.multi-tenancy.strategies';
