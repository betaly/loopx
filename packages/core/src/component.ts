import {HelmetComponent} from '@bleco/helmet';
import {RateLimiterComponent} from '@bleco/ratelimiter';
import {Binding, Component, CoreBindings, createBindingFromClass, inject, ProviderMap} from '@loopback/core';
import {ExpressRequestHandler, RestApplication, RestBindings} from '@loopback/rest';
import {configure} from 'i18n';
import * as swstats from 'swagger-stats';

import {LoggerExtensionComponent, SwaggerAuthenticationComponent} from './components';
import {OperationSpecEnhancer} from './enhancer/operation-spec-enhancer';
import {LocaleKey} from './enums';
import {LxCoreBindings, OASBindings} from './keys';
import {LxCoreConfig} from './types';

export class LxCoreComponent implements Component {
  localeObj: i18nAPI = {} as i18nAPI;
  providers?: ProviderMap = {};
  bindings: Binding[] = [];

  constructor(
    @inject(CoreBindings.APPLICATION_INSTANCE)
    private readonly application: RestApplication,
    @inject(LxCoreBindings.config, {optional: true})
    private readonly coreOptions: LxCoreConfig,
    @inject(LxCoreBindings.EXPRESS_MIDDLEWARES, {optional: true})
    private readonly expressMiddlewares: ExpressRequestHandler[],
  ) {
    const middlewares = [];
    if (this.expressMiddlewares) {
      middlewares.push(...this.expressMiddlewares);
    }

    // Mount logger component
    this.application.component(LoggerExtensionComponent);

    this.application.component(HelmetComponent);
    this.application.component(RateLimiterComponent);

    // Enable OBF
    if (this.coreOptions?.enableObf && this.coreOptions?.openapiSpec) {
      const middlewareConfig = Object.assign(this.coreOptions.swaggerStatsConfig ?? {}, {
        name: this.coreOptions?.name,
        uriPath: this.coreOptions?.obfPath ?? `/obf`,
        swaggerSpec: this.coreOptions?.openapiSpec,
        authentication: this.coreOptions.authentication ?? false,
      });
      const swStatsMiddleware = swstats.getMiddleware({
        ...middlewareConfig,
        onAuthenticate: this.coreOptions.swaggerAuthenticate
          ? this.coreOptions.swaggerAuthenticate
          : (req, username, password) => {
              return username === this.coreOptions.swaggerUsername && password === this.coreOptions.swaggerPassword;
            },
      });
      middlewares.push(swStatsMiddleware);
    }

    if (this.coreOptions?.authenticateSwaggerUI) {
      this.application.component(SwaggerAuthenticationComponent);
    }

    // Configure locale provider
    if (this.coreOptions?.configObject) {
      configure({...this.coreOptions.configObject, register: this.localeObj});
    } else {
      configure({
        locales: [LocaleKey.en, LocaleKey.es, LocaleKey.ptBr, LocaleKey.ptPt, LocaleKey.esCo],
        fallbacks: {
          [LocaleKey.es]: 'en',
          [LocaleKey.esCo]: 'en',
          [LocaleKey.ptBr]: 'en',
          [LocaleKey.ptPt]: 'en',
        },
        register: this.localeObj,
        directoryPermissions: '777',
        directory: `${__dirname}/../locales`,
        // sonarignore:start
        /* eslint-disable @typescript-eslint/no-explicit-any */
        objectNotation: '->' as any,
        // sonarignore:end
      });
    }

    // Configure error writer options
    if (!this.application.isBound(RestBindings.ERROR_WRITER_OPTIONS)) {
      this.application.bind(RestBindings.ERROR_WRITER_OPTIONS).to({
        safeFields: ['code', 'errorCode', 'data'],
      });
    }

    this.application.bind(LxCoreBindings.EXPRESS_MIDDLEWARES).to(middlewares);
    this.bindings.push(Binding.bind(OASBindings.HiddenEndpoint).to([]));
    this.bindings.push(Binding.bind(LxCoreBindings.i18n).to(this.localeObj));
    this.application.add(createBindingFromClass(OperationSpecEnhancer));
    this.application.bind(LxCoreBindings.DEFAULT_TENANT_KEY).to(this.coreOptions?.defaultTenantKey ?? 'default');
  }
}
