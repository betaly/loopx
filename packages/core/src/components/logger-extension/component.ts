import {Component, ProviderMap} from '@loopback/core';

import {LOGGER} from './keys';
import {PinoLoggerProvider} from './providers/pino-logger.provider';

export class LoggerExtensionComponent implements Component {
  providers: ProviderMap = {};
  constructor() {
    this.providers = {[LOGGER.BINDINGS.LOG_ACTION.key]: PinoLoggerProvider};
  }
}
