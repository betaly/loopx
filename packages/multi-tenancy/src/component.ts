import {Binding, Component, Constructor, createBindingFromClass, extensionFor} from '@loopback/core';

import {MULTI_TENANCY_STRATEGIES, MultiTenancyBindings} from './keys';
import {MultiTenancyMiddlewareProvider} from './middleware/multi-tenancy-middleware.provider';
import {DefaultTenantResolverProvider} from './providers';
import {strategies} from './strategies';

export class MultiTenancyComponent implements Component {
  bindings: Binding[] = [
    createBindingFromClass(MultiTenancyMiddlewareProvider, {
      key: MultiTenancyBindings.MIDDLEWARE,
    }),
    ...strategies.map((strategy: Constructor<unknown>) =>
      createBindingFromClass(strategy).apply(extensionFor(MULTI_TENANCY_STRATEGIES)),
    ),
  ];

  providers = {
    [MultiTenancyBindings.TENANT_RESOLVER.key]: DefaultTenantResolverProvider,
  };
}
