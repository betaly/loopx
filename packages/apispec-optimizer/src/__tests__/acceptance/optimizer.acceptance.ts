import {Application} from '@loopback/core';
import {Entity, model, property} from '@loopback/repository';
import {get, param, RestApplication, RestServer} from '@loopback/rest';

import {ApiSpecOptimizerBindings} from '../../keys';
import {ApiSpecOptimizerComponent} from '../../optimizer.component';
import {ApiSpecOptimizerOptions} from '../../optimizer.spec-enhancer';

describe('Extension for API spec optimizer - OASEnhancer', () => {
  let app: Application;
  let server: RestServer;
  beforeEach(givenAServer);

  @model()
  class Todo extends Entity {
    @property({
      type: 'number',
      id: true,
      generated: false,
    })
    id: number;
  }

  class TodoController {
    @get('/todos/{id}')
    async findById(@param.path.number('id') id: number): Promise<Todo> {
      return new Todo();
    }

    @get('/todos')
    async find(): Promise<Todo[]> {
      return [new Todo()];
    }

    @get('/todos/count')
    async count(): Promise<number> {
      return 1;
    }
  }

  class PingController {
    @get('/ping')
    async ping(): Promise<string> {
      return 'pong';
    }
  }

  class ThisIsMyWorldWordController {
    @get('/my-world')
    async find(): Promise<string> {
      return 'hello world';
    }
  }

  it('optimize api spec', async () => {
    const spec = await server.getApiSpec();
    expect(spec).toMatchSnapshot();
  });

  async function givenAServer() {
    app = new RestApplication();
    app.component(ApiSpecOptimizerComponent);
    const apiSpecOptimizerOptions: ApiSpecOptimizerOptions = {
      // customizeControllerName: name => name,
      // customizeOperationId: (name: string, op: OperationObject, spec: OpenApiSpec) => name,
    };
    app.configure(ApiSpecOptimizerBindings.API_SPEC_OPTIMIZER).to(apiSpecOptimizerOptions);
    app.controller(TodoController);
    app.controller(PingController);
    app.controller(ThisIsMyWorldWordController);
    server = await app.getServer(RestServer);
  }
});
