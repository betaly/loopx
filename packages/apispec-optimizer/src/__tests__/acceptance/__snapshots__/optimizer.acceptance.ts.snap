// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension for API spec optimizer - OASEnhancer optimize api spec 1`] = `
{
  "info": {
    "title": "LoopBack Application",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/my-world": {
      "get": {
        "operationId": "findThisIsMyWorldWords",
        "responses": {
          "200": {
            "description": "Return value of ThisIsMyWorldWordController.find",
          },
        },
        "tags": [
          "ThisIsMyWorldWord",
        ],
        "x-controller-name": "ThisIsMyWorldWord",
        "x-operation-name": "find",
      },
    },
    "/ping": {
      "get": {
        "operationId": "ping",
        "responses": {
          "200": {
            "description": "Return value of PingController.ping",
          },
        },
        "tags": [
          "Ping",
        ],
        "x-controller-name": "Ping",
        "x-operation-name": "ping",
      },
    },
    "/todos": {
      "get": {
        "operationId": "findTodos",
        "responses": {
          "200": {
            "description": "Return value of TodoController.find",
          },
        },
        "tags": [
          "Todo",
        ],
        "x-controller-name": "Todo",
        "x-operation-name": "find",
      },
    },
    "/todos/count": {
      "get": {
        "operationId": "countTodos",
        "responses": {
          "200": {
            "description": "Return value of TodoController.count",
          },
        },
        "tags": [
          "Todo",
        ],
        "x-controller-name": "Todo",
        "x-operation-name": "count",
      },
    },
    "/todos/{id}": {
      "get": {
        "operationId": "findTodoById",
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "number",
            },
          },
        ],
        "responses": {
          "200": {
            "description": "Return value of TodoController.findById",
          },
        },
        "tags": [
          "Todo",
        ],
        "x-controller-name": "Todo",
        "x-operation-name": "findById",
      },
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;
